### ISSUE TITLE
Filetype seems detected by filetype.lua but is not applied

### ISSUE THREAD

**Issue by gautaz** [neovim/neovim#23522]
### Problem

Hello,

This is most probably a PEBKAC but as I did not found any issue or post related to this behavior, I am opening this issue as a last resort.

I have the following line in my `init.lua`:

```lua
vim.filetype.add({ extension = { nosj = "json" } })
```

So I am expecting `neovim` to display all files with a `.nosj` extension as a JSON file.

Currently it does not (at least on my workstation).
For instance, opening a file named `file.nosj` and issuing `:echo &filetype` shows nothing (issuing the same command on a file named `file.json` shows `json`).

But issuing `:lua print(vim.filetype.match({ filename = 'file.nosj' }))` shows `json nil` which is the same result as `:lua print(vim.filetype.match({ filename = 'file.json' }))` so `filetype.lua` seems to detect the extension as expected.

### Steps to reproduce

Here is the minimal configuration reproducing the issue (`issue.lua`):
```lua
vim.filetype.add({ extension = { nosj = "json" } })
```

```sh
nvim --clean -u issue.lua
:lua print(vim.filetype.match({ filename = 'file.nosj' })) # shows "json nil"
:e file.nosj
:echo &filetype # shows nothing
:e file.json
:echo &filetype # shows "json"
```

### Expected behavior

```sh
nvim --clean -u issue.lua
:e file.nosj
:echo &filetype # shows "json"
```

### Neovim version (nvim -v)

NVIM v0.9.0

### Vim (not Nvim) behaves the same?

No, filetype.lua is nvim specific

### Operating system/version

Linux hepao 6.1.27 #1-NixOS SMP PREEMPT_DYNAMIC Sun Apr 30 23:26:28 UTC 2023 x86_64 GNU/Linux

### Terminal name/version

alacritty 0.12.0

### $TERM environment variable

tmux-256color

### Installation

Nix Home Manager

**Comment by clason** [comment]
Works for me. Can you try the latest nightly, just to be sure?

**Comment by gautaz** [comment]
Hello @clason, I have just tried with the latest sources (commit `1cbfed03c249e7f9e67d59566fbabe46f7f7f1f9`, `nvim -v` reports `NVIM v0.10.0-dev`).

I still have the same behavior...
Should `vim.filetype.add({ extension = { nosj = "json" } })` be enough for the filetype to be set accordingly?

**Comment by clason** [comment]
It is, for me.

**Comment by gautaz** [comment]
Just for the record, I have tried with the latest source but still by using a Nix generation based on `pkgs.neovim-unwrapped` (see https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/programs/neovim.nix#L91).

Trying to build directly from a `git clone` without relying on a Nix recipe can be quite painful under NixOS (I tried but I almost immediately stumbled over issues for finding libs scattered all over the Nix store).

**Comment by gautaz** [comment]
As the initial analysis seems to point a `nixpkgs` specific issue, I have opened a new issue on `NisOS/nixpkgs` related to this issue.

**Comment by amerry** [comment]
I'm seeing a similar thing on Windows (0.9.0, installed via scoop): I have

```
lua vim.filetype.add { extension = { jsonschema = 'json' } }
```

in my `init.vim`, but my `<whatever>.jsonschema` files are not getting json highlighting (but if I run `:setf json` in the buffer, that works).

**Comment by ryuheechul** [comment]
_adding my case very similar to this issue (probably the same)_

I also experience the same kind of issue ever since sometime after 0.9. Also not only `vim.filetype.add` doesn't work but also the traditional way as well, [https://github.com/ryuheechul/dotfiles/blob/[historical master]/nvim/ftdetect/detect.vim](https://github.com/ryuheechul/dotfiles/blob/078f2e5a781050a3e1655b03684fd09d86f6640a/nvim/ftdetect/detect.vim).

I do install neovim via nixpkgs (via home-manager) and this issue occurs both on NixOS and macOS.

Only the file detection is the issue not highlighting and such.
(`set ft?` result in `filetype=`)

Currently working around this by adding a comment on the top of each file `// vim: set ft=kdl:`.

**Comment by 9999years** [comment]
Seeing this with Neovim 0.9.4, both the Homebrew and the Nixpkgs versions.

**Comment by gemelen** [comment]
Adding to this, the same "empty" behaviour observed both for setting filetype via `extension` and `pattern`.
Setup - 0.9.5, regular release build on openSuSE.

**Comment by ryuheechul** [comment]
I have a new lead, folks.
Check if you are using [vim-polyglot](https://github.com/sheerun/vim-polyglot). After removing `vim-polyglot` plugin, all kinds of weird silent errors has been disappeared including this issue.
(and the issue reappear as soon as I re-enable `vim-polyglot`)

**Comment by gautaz** [comment]
Hello @ryuheechul,

Well done, good finding!
I can confirm that removing vim-polyglot fixes the issue on my side.
So the next step is to open an issue on vim-polyglot.

@ryuheechul As you found the culprit, do you want to open the issue?
I can open it if you are in a hurry.

**Comment by gemelen** [comment]
Sorry, but I don't think it's the `vim-polyglot` on its own, but rather one of its component's behaviour. I could tell because I don't use this plugin, but observe the problematic behaviour too.

**Comment by ryuheechul** [comment]
@gautaz yeah finally the mystery is spotted! (kind of)
I decided not to use `vim-polyglot` as of yesterday after the discovery of multiple (also other) issues occurring with `vim-polygot` being together with the rest of my Neovim config and I decided that it's will be less headache to just all in on Treesitter (especially due to my current particular environment is influencing weird conditional poor regex performance in Neovim which is being arbitrarily documented [here](https://github.com/ryuheechul/dotfiles/blob/master/nvim/lua/plugins/README.md)).

But this is my personal choice and others probably would pursue different paths. 
I also agree with @gemelen that we still don't know what exactly causing this issue, I'm looking forward to more findings.

**Comment by merikan** [comment]
Awesome, kudos to you @ryuheechul 🙏
This has been bugging me for a long time now and I can also confirm that after disabling the polyglot plugin my `.kdl`, `.templ` files are now detected.

**Comment by gautaz** [comment]
> I also agree with @gemelen that we still don't know what exactly causing this issue, I'm looking forward to more findings.

Yes, I double this, @gemelen, would you mind sharing your list of vim plugins so we can find what it has in common with the list of plugins integrated by `vim-polyglot`?

@merikan finding about the `ftdetect` setting may also lead to the right track.

**Comment by justinmk** [comment]
No repro, not actionable.

**Comment by gautaz** [comment]
Hello @justinmk, I just tried to reproduce the issue with my current neovim version (v0.11.6) and it is in fact fixed now.
Thanks for closing the issue.
